nomad/client_agent_endpoint.go

```diff
@@ -23,11 +24,95 @@ type Agent struct {
 	srv *Server
 }
 
-func (m *Agent) register() {
-	m.srv.streamingRpcs.Register("Agent.Monitor", m.monitor)
+func (a *Agent) register() {
+	a.srv.streamingRpcs.Register("Agent.Monitor", a.monitor)
 }
 
-func (m *Agent) monitor(conn io.ReadWriteCloser) {
+func (a *Agent) Profile(args *structs.AgentPprofRequest, reply *structs.AgentPprofResponse) error {
+	// Forward to different region if necessary
+	// this would typically be done in a.srv.forward() but since
+	// we are targeting a specific server, not just the leader
+	// we must manually handle region forwarding here.
+	region := args.RequestRegion()
+	if region == "" {
+		return fmt.Errorf("missing target RPC")
+	}
+
+	if region != a.srv.config.Region {
+		// Mark that we are forwarding
+		args.SetForwarded()
+		return a.srv.forwardRegion(region, "Agent.Profile", args, reply)
+	}
+
+	// Targeting a node, forward request to node
+	if args.NodeID != "" {
+		return a.forwardProfileClient(args, reply)
+	}
+
+	// Handle serverID not equal to ours
+	if args.ServerID != "" {
+		serverToFwd, err := a.forwardFor(args.ServerID, region)
+		if err != nil {
+			return err
+		}
+		if serverToFwd != nil {
+			return a.srv.forwardServer(serverToFwd, "Agent.Profile", args, reply)
+		}
+	}
+
+	// Check ACL for agent write
+	aclObj, err := a.srv.ResolveToken(args.AuthToken)
+	if err != nil {
+		return err
+	} else if aclObj != nil && !aclObj.AllowAgentWrite() {
+		return structs.ErrPermissionDenied
+	}
```

Comment:
Should we optimistically move this above client/server forwarding conditionals?  The pattern in other functions is to avoid forwarding if we confirm request isn't authenticated.  Given that `EnableDebug` is a agent specific, keeping it as-is seems good?

Reply:
yeah the thought was to keep the debug and acl check close together, but we could move the acl bit above the forwarding

Reply:
(moving up)